bridges/snowbridge/parachain/pallets/inbound-queue/src/lib.rs

```diff
@@ -0,0 +1,330 @@
+				XcmpSendError::DestinationUnsupported =>
+					Error::<T>::Send(SendError::DestinationUnsupported),
+				XcmpSendError::ExceedsMaxMessageSize =>
+					Error::<T>::Send(SendError::ExceedsMaxMessageSize),
+				XcmpSendError::MissingArgument => Error::<T>::Send(SendError::MissingArgument),
+				XcmpSendError::Fees => Error::<T>::Send(SendError::Fees),
+			}
+		}
+	}
+
+	/// The current nonce for each channel
+	#[pallet::storage]
+	pub type Nonce<T: Config> = StorageMap<_, Twox64Concat, ChannelId, u64, ValueQuery>;
+
+	/// The current operating mode of the pallet.
+	#[pallet::storage]
+	#[pallet::getter(fn operating_mode)]
+	pub type OperatingMode<T: Config> = StorageValue<_, BasicOperatingMode, ValueQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Submit an inbound message originating from the Gateway contract on Ethereum
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::submit())]
+		pub fn submit(origin: OriginFor<T>, message: Message) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			ensure!(!Self::operating_mode().is_halted(), Error::<T>::Halted);
+
+			// submit message to verifier for verification
+			T::Verifier::verify(&message.event_log, &message.proof)
+				.map_err(|e| Error::<T>::Verification(e))?;
+
+			// Decode event log into an Envelope
+			let envelope =
+				Envelope::try_from(message.event_log).map_err(|_| Error::<T>::InvalidEnvelope)?;
+
+			// Verify that the message was submitted from the known Gateway contract
+			ensure!(T::GatewayAddress::get() == envelope.gateway, Error::<T>::InvalidGateway);
+
+			// Retrieve the registered channel for this message
+			let channel =
+				T::ChannelLookup::lookup(envelope.channel_id).ok_or(Error::<T>::InvalidChannel)?;
+
+			// Verify message nonce
+			<Nonce<T>>::try_mutate(envelope.channel_id, |nonce| -> DispatchResult {
+				if *nonce == u64::MAX {
+					return Err(Error::<T>::MaxNonceReached.into())
+				}
+				if envelope.nonce != nonce.saturating_add(1) {
+					Err(Error::<T>::InvalidNonce.into())
+				} else {
+					*nonce = nonce.saturating_add(1);
+					Ok(())
+				}
+			})?;
+
+			// Reward relayer from the sovereign account of the destination parachain
+			// Expected to fail if sovereign account has no funds
+			let sovereign_account = sibling_sovereign_account::<T>(channel.para_id);
+			let delivery_cost = Self::get();
+			T::Token::transfer(&sovereign_account, &who, delivery_cost, Preservation::Preserve)?;
+
+			// Decode message into XCM
+			let (xcm, fee) =
+				match inbound::VersionedMessage::decode_all(&mut envelope.payload.as_ref()) {
+					Ok(message) => Self::do_convert(envelope.message_id, message)?,
+					Err(_) => return Err(Error::<T>::InvalidPayload.into()),
+				};
+
+			// We embed fees for xcm execution inside the xcm program using teleports
+			// so we must burn the amount of the fee embedded into the XCM script.
+			T::Token::burn_from(&sovereign_account, fee, Precision::Exact, Fortitude::Polite)?;
```

Comment:
Just to confirm:
1) sender pays this fee at Eth side and funds go to account, associated with BH (bridge hub), right? Or are they burnt there? (sorry - I haven't been reading contract details there)
2) we burn this fee from AH (asset hub) account here;
3) at AH this BH receives the `fee` and pays for the execution and receives the surplus.

So (2) + (3) means that the AH is pays for the XCM execution.

It caught my eye, because if fees at step (1) go to the account, associated with BH (which I'm not sure), then why the AH pays for the execution here?

Reply:
In step 1, the collected fees (ether) go into an ethereum contract controlled by AssetHub (and by extension, polkadot governance).

> So (2) + (3) means that the AH is pays for the XCM execution.

Correct, as AssetHub's sovereign account on Ethereum is being credited with collected fees. So the debits and credits should balance.

I've documented how the cross-chain fee system works here: https://docs.snowbridge.network/architecture/fees-and-channels

Reply:
Cool, thank you for explaining!